### PR TITLE
WIP Adding position and tilt to template cover

### LINKED ISSUE #42: Template cover SET_COVER_POSITION support
Do you have any plans to support [setting cover position](https://www.home-assistant.io/components/cover/#service-coverset_cover_position) for template covers? Additionally, is there a good way to tie a cover position sensor to the position of a slider in the Home Assistant UI? It seems like existing options require the sensor/slider to be pretty much separate from the cover and its open/closed state and open/stop/close buttons.

### PR DESCRIPTION
**Pull request by mtl010957**
## Description:
This is my initial working version of adding position and tilt setting capabilities to the existing template cover component. This version works in optimistic mode but I think it is not working properly with lambdas to generate position and tilt states. Please review and set me on the right path.

**Related issue (if applicable):** fixes [feature-request #42](https://github.com/esphome/feature-requests/issues/42)

**Pull request in [esphome](https://github.com/esphome/esphome) with python changes (if applicable):** esphome/esphome#<esphome PR number goes here>
**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:

  - [X] The code change is tested and works locally.
  - [X] The code change follows the [standards](https://esphome.io/guides/contributing.html#contributing-to-esphome-core)

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: src/esphome/cover/cover.h [hunk 1]
```diff
@@ -22,12 +22,21 @@ enum CoverCommand {
   COVER_COMMAND_OPEN = 0,
   COVER_COMMAND_CLOSE,
   COVER_COMMAND_STOP,
+  COVER_COMMAND_POSITION,
+  COVER_COMMAND_TILT,
 };
 
+using position_value_t = float;
+using tilt_value_t = float;
```
Comment (OttoWinter) [inline src/esphome/cover/cover.h]: I think we should drop these typedefs. They're nice, but in C++ there's no way to enforce the use of them and you end up with something where everything is fragmented in parts where the typedef is used and parts where the base type is used directly. For example, with `position_closed_` etc below you use floats even though these typedefs exist, same goes for write_command.

**Review by OttoWinter — APPROVED** [review]
There's a problem with the MQTT side: The MQTT side will subscribe to the tilt/position topics even if the cover does not support those things.

Also, the implementation here feels a lot like it was bolted on top of the old design, but the old design was never meant for this. For example `write_command` should not be used for position/tilt. That should be a separate method.

So this PR needs:
- A "traits" model for covers. Cover integrations should override an abstract method `get_traits` that informs what things a cover supports.
- `write_command` should be removed and split into multiple methods, or best a `CoverCall` class should be added like for the climate base (see climate branch)

I can work on these two things later, I also want to see this supported with the next release (got a few covers myself lately, and they support tilt)

File: src/esphome/cover/mqtt_cover_component.cpp [hunk 2]
```diff
@@ -28,19 +28,133 @@ void MQTTCoverComponent::setup() {
...
+    ESP_LOGCONFIG(TAG, "  Position State Topic '%s':", this->custom_position_state_topic_.c_str());
+  if (!this->custom_tilt_command_topic_.empty())
+    ESP_LOGCONFIG(TAG, "  Tilt Command Topic '%s':", this->custom_tilt_command_topic_.c_str());
+  if (!this->custom_tilt_status_topic_.empty())
+    ESP_LOGCONFIG(TAG, "  Tilt Status Topic '%s':", this->custom_tilt_status_topic_.c_str());
 }
 void MQTTCoverComponent::send_discovery(JsonObject &root, mqtt::SendDiscoveryConfig &config) {
+  if (this->custom_command_topic_.empty()) {
+//    config.state_topic = false;
+    config.command_topic = false;
```
[the remark this answers is not in the document]
Comment (mtl010957) [reply src/esphome/cover/mqtt_cover_component.cpp:84]: This setting of the command topic configuration flag to false is needed to turn off the basic state behavior if the cover has only tilt and/or position and no explicit position command. Maybe I didn't understand how that combination was intended to work, but in my case I use tilt only and the tilt position in relation to the tilt_opened_position and tilt_closed_position values determines the state of the cover.

**Comment by mtl010957** [comment]
@OttoWinter Thanks so much for all the help with this, I'm still ramping up my understanding of your architecture and how C++ handles many things so this is very helpful. I'll be working this one slowly over the next few weeks so please take over and drive this the right direction if you want and I'll try to follow and help.

**Comment by mtl010957** [comment]
Closing this PR, the whole approach was a bit flawed. Due to the refactor all of the new work on position and tilt covers is now in esphome dev branch.

**Pull request closed without merging**